Debug-frontend-tests.md

```diff
@@ -0,0 +1,98 @@
+      ...
+    });
+  });
+});
+```
+
+The strings from the `describe` and `it` functions get concatenated with space (` `) delimiters to create the test name. The name of the `it` block usually starts with `should`, so we recommend searching the codebase for the portion of the test name from `should` to the end (e.g. "should handle rejection when fetching a file fails" in this case).
+
+## Isolate tests
+
+Once you have identified which test is failing, you often want to run that test in isolation next. This will let you get results quickly as you try other debugging strategies. Thankfully, you just have to change the `it()` of your test to `fit()`. You can even do this multiple times to mark multiple tests for running. Then when you run the frontend tests, only those tests with `fit()` will run. Note that `fdescribe()` works the same way.
+
+**Remember to revert your `fit()` and `fdescribe()` changes before committing!**
+
+## Verbose mode
+
+By default, we suppress log output from the tests to the terminal when running the frontend tests. This keeps the test output clean so it's easy to find which tests failed. However, this output can be useful for debugging. For example, if you add `console.log()` or `console.error()` statements to your tests for debugging, the output from those statements will be suppressed by default. To view this output, pass `--verbose` when you run the frontend tests:
+
+```console
+python -m scripts.run_frontend_tests --verbose
+```
+
+**Only use --verbose when you are running a few tests in isolation. Otherwise, you will be swamped with way too many log messages.**
+
+Note that depending on your situation, `console.log()` or `console.error()` might be preferable:
+
+* `console.log()` statements will cause the linter to fail, so they are a great choice for local debugging. Then if you forget to remove them before pushing your changes, the linter will remind you.
+* `console.error()` statements do not cause a lint failure, so they work well when you are pushing your code with debugging code to a PR to let the tests run in the CI environment.
+
+## Find stack elements
+
+While you can't use the line numbers in a stack trace to find the associated code, you can sometimes use the function or class names. For example, in the example above we could try searching the codebase for `_maybeConvertBody`. The success of this technique depends on how frequently the string you search for occurs in the codebase. If the string occurs a lot, you will have a hard time figuring out which occurrence is the one you are interested in.
+
+## Run tests repeatedly on CI
+
+Especially when debugging flaky frontend tests, you may want to set your tests to run many times on CI. For example, you might add some debugging code and then rerun your frontend tests on CI until the flake occurs. You can update the frontend test CI workflow to enable this. In `.github/workflows/frontend_tests.yml`, change the step that runs the tests to:
```

Comment:
Do we recommend them to run the GitHub action on their own fork? Or is it okay to run on the Oppia repo itself?

Reply:
We want them to use their own fork. I've added that